sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala

```diff
@@ -304,8 +504,196 @@ class RapidsBufferCatalog extends AutoCloseable with Arm with Logging {
+      withResource(buffer) { _ =>
+        logWarning(s"Spilling $buffer ${buffer.id} to ${spillStore.name}")
+        val bufferHasSpilled = isBufferSpilled(buffer.id, buffer.storageTier)
+        if (!bufferHasSpilled) {
+          val spillCallback = buffer.getSpillCallback
+          spillCallback(buffer.storageTier, spillStore.tier, buffer.size)
+
+          // if the spillStore requires it, make room for buffer.
+          makeRoomForBufferIfNeeded(buffer, spillStore)
+
+          // copy the buffer to spillStore
+          var newBuffer: Option[RapidsBuffer] = None
+          while (newBuffer.isEmpty) {
+            newBuffer = spillStore.copyBuffer(buffer, buffer.getMemoryBuffer, stream)
+            if (newBuffer.isEmpty) {
+              // only the host store will return None above. We assume that the allocate function
+              // from `HostMemoryBuffer` throws and breaks this loop if we run out of host memory
+              synchronousSpill(spillStore, math.max(spillStore.currentSize - buffer.size, 0))
+            }
+          }
+
+          // once spilled, we get back a new RapidsBuffer instance in this new tier
+          registerNewBuffer(newBuffer.get)
+        } else {
+          logDebug(s"Skipping spilling $buffer ${buffer.id} to ${spillStore.name} as it is " +
+            s"already stored in multiple tiers")
+        }
+      }
+      // we can now remove the old tier linkage
+      removeBufferTier(buffer.id, buffer.storageTier)
+      // and free
+      buffer.safeFree()
+    }
+  }
+
+  /**
+   * If `spillStore` defines a maximum size, make room for `buffer` by spilling.
+   */
+  private def makeRoomForBufferIfNeeded(
+      buffer: RapidsBuffer,
+      spillStore: RapidsBufferStore): RapidsBuffer = {
+    val spillStoreMaxSize = spillStore.getMaxSize
+    if (spillStoreMaxSize.isDefined) {
+      // this spillStore has a maximum size requirement (host only). We need to spill from it
+      // in order to make room for `buffer`.
+      val targetTotalSize =
+        math.max(spillStoreMaxSize.get - buffer.size, 0)
+      val maybeAmountSpilled = synchronousSpill(spillStore, targetTotalSize)
+      maybeAmountSpilled.foreach { amountSpilled =>
+        if (amountSpilled != 0) {
+          logInfo(s"Spilled $amountSpilled bytes from the ${spillStore.name} store")
+          TrampolineUtil.incTaskMetricsDiskBytesSpilled(amountSpilled)
+        }
+      }
+    }
+  }
+
+  /**
+   * Copies `buffer` to the `deviceStorage` store, registering a new `RapidsBuffer` in
+   * the process
+   * @param buffer - buffer to copy
+   * @param memoryBuffer - cuDF MemoryBuffer to copy from
+   * @param stream - Cuda.Stream to synchronize on
+   * @return - The `RapidsBuffer` instance that was added to the device store.
+   */
+  def unspillBufferToDeviceStore(
+    buffer: RapidsBuffer,
+    memoryBuffer: MemoryBuffer,
+    stream: Cuda.Stream): RapidsBuffer = synchronized {
+    val newBuffer = deviceStorage.copyBuffer(
```

Comment:
make sure two threads calling here won't re-materialize the buffer into device memory.

Reply:
this needs to return an old RapidsBuffer if it already existed in the device tier.